analyzers/src/SonarAnalyzer.CSharp/Rules/UseCachedRegex.cs

```diff
@@ -0,0 +1,114 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Rules.CSharp;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public sealed class UseCachedRegex : UseCachedRegexBase<SyntaxKind>
+{
+    protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+    protected override void Initialize(SonarAnalysisContext context) =>
+        context.RegisterNodeAction(c =>
+            {
+                var objectCreationExpression = (ObjectCreationExpressionSyntax)c.Node;
+
+                if (objectCreationExpression.NameIs(nameof(Regex))
+                    && IsWithinCorrectContext(objectCreationExpression)
+                    && IsCorrectObjectType(objectCreationExpression, c.SemanticModel)
+                    && IsArgumentConstantOrReadOnly(objectCreationExpression, c.SemanticModel)
+                    && !IsCompliantAssignment(objectCreationExpression, c.SemanticModel))
+                {
+                    c.ReportIssue(Diagnostic.Create(Rule, objectCreationExpression.GetLocation()));
+                }
+            },
+            SyntaxKind.ObjectCreationExpression);
+
+    private static bool IsCompliantAssignment(ObjectCreationExpressionSyntax objectCreationExpression, SemanticModel model) =>
+        objectCreationExpression.Parent switch
+        {
+            AssignmentExpressionSyntax { RawKind: (int)SyntaxKind.SimpleAssignmentExpression } assignmentExpression =>
```

Comment:
(Optional)
to avoid nesting and also make it a bit more readable, you can use `when` pattern-matching:

```csharp
AssignmentExpressionSyntax assignment when assignment.IsKind(SyntaxKind.SimpleAssignment) =>
```